glossary.Rmd

```diff
@@ -0,0 +1,163 @@
+
+```{r}
+base::expression(key <- "foo", toupper(key))
+```
+
+You'll most likely encounter this rare data structure as the return value of `base::parse()`:
+
+```{r}
+code <- "key <- 'foo'; toupper(key)"
+parse(text = code)
+```
+
+The only advantage of expression vectors compared to lists is that they include **source references**. The most apparent manifestation of source references is that expression vectors are printed with whitespace and comments preserved:
+
+```{r}
+code <- "{
+  # Interesting comment
+  weird     <-     whitespace
+}"
+parse(text = code, keep.source = TRUE)
+```
+
+Source references are mostly useful for debugging and development tools. They don't play any computational role and tidy evaluation doesn't make use of references. Consequently the parsing tools in rlang return normal lists of expressions (in the rlang sense) instead of expression vectors:
+
+```{r}
+rlang::parse_exprs(code)
+```
+
+
+### TODO Symbol {#glossary-sym}
+
+
+## Programming Concepts
+
+### Constant versus symbolic {#glossary-constant-symbolic}
+
+Constants, also called "literals", always have the same value no matter the context. On the other hand symbols and calls are [symbolic](#glossary-constant-symbolic) objects: their value depends on an [environment](#glossary-env) and what kind of objects are defined there.
+
+For instance the string `"mickey"` always represents the same string no matter the environment and what objects are defined there:
+
+```{r}
+# Here's a string:
+"mickey"
+
+mickey <- "mouse"
+
+# Still the same string:
+"mickey"
+```
+
+On the other [symbols](#glossary-sym) depend on current definitions:
+
+```{r}
+# We've defined `mickey` as "mouse"
+mickey
+
+mickey <- "mickey"
+
+# Now `mickey` is "mickey"
+mickey
+```
+
+One source of problem when you're working with quoted expressions is that they might be evaluated in arbitrary places, where objects have potentially been redefined to something completely different than expected. This is a common issue with tidyverse grammars because they evaluate quoted expressions in a [data mask](#glossary-data-mask). Say you'd like to divide a column by a factor defined in the current environment:
+
+```{r}
+factor <- 100
+
+starwars %>% mutate(height / factor) %>% pull()
+```
+
+This works fine but what if the data frame contains a column called `factor`? The expression will be evaluated with the wrong definition:
+
+```{r}
+# Derive a data frame that contains a `factor` column
+starwars2 <- starwars %>% mutate(factor = 1:n())
+
+# Oh no! We're now dividing `height` by the new column!
+starwars2 %>% mutate(height / factor) %>% pull()
+```
+
+Masking is generally not a problem in scripts because you know what columns are inside your data frame. However as soon as your code is getting more general, for instance if you create a reusable function, you can no longer make assumptions about what's in the data.
```

Comment:
I think this is really important, because it's a big piece of why tidy eval is important for some people and not for others (no change to content, but we might consider moving the "problem scenario" earlier).

Reply:
I think this will move to an actual section of the bookdown in the future, and we'll just link to this section.

Reply:
Yeah. I'm realising I may be insufficiently treating this as a glossary—glossary definitions are usually self-contained, but don't necessarily put the pieces together for you (which is what the actual text of a book does). So, summary point: that sounds good.